lxd/instance/drivers/driver_qemu.go

```diff
@@ -1588,8 +1588,11 @@ func (vm *qemu) generateQemuConfigFile(busName string, devConfs []*deviceConfig.
 	// Setup the bus allocator.
 	bus := qemuNewBus(busName, sb)
 
-	// Now add the fixed set of devices.
-	devBus, devAddr, multi := bus.allocate("generic")
+	// Now add the fixed set of devices. The multi-function groups used for these fixed internal devices are
+	// specifically chosen to ensure that we consume exactly 4 PCI bus ports (on PCIe bus). This ensures that
```

Comment:
@stgraber I thought adding a comment about why the specific bus function groups have been picked here would be useful in the future, although I'm not sure I've got the reasoning 100% correct given the different architectures we support.

Reply:
It's also meant to group all low-bandwidth internal devices onto a single address.

We can plug a total of 256 things into a PCIe bus, but this assumes 32 chassis * 8 function. If we waste chassis with internal devices, we'll end up seriously limiting the number of network interfaces or vfio devices that can be attached as those need a full chassis for themselves.

Reply:
@stgraber thats great, will update comment to save knowledge/design reasoning.